docker_parallel_runner.py

```diff
@@ -0,0 +1,320 @@
+    # We leverage the ThreadPoolExecutor for its simple threading interface
+    # Therefore we effectively do not constrain on max_workers (set to theoretical max usage == len(suites))
+    executor = ThreadPoolExecutor(max_workers=len(suites))
+
+    actual_cpu_available = os.cpu_count()
+    assert actual_cpu_available, "Could not determine available cpu count on system"
+    logging.info(f"Actual total CPU available on system: {actual_cpu_available}")
+    total_cpu_available = float(actual_cpu_available * core_count_mult)
+    current_cpu_available = total_cpu_available
+    logging.info(f"Effective CPU available (core_count_mult={core_count_mult}): {current_cpu_available}")
+
+    running_futures: List[Future] = []
+    succeeded_count = 0
+    error_count = 0
+
+    # sort suites by CPU requirement, descending
+    suites_pending_submission = sorted(suites, key=lambda s: s.cpus_needed, reverse=True)
+
+    # Keep going, if either there are things to submit, or things are running (and need to be waited on)
+    while suites_pending_submission or running_futures:
+        if error_count > 0:
+            logging.error(f"SOMETHING ERROR'D OUT - NOT STARTING ANY MORE SUITES (remaining: {suites_pending_submission})")
+            break
+
+        # eagerly submit as many suites as possible
+        # chunkiest first, no lookahead
+        submitted_suites: List[Suite] = []
+        for s in suites_pending_submission:
+            if s.cpus_needed <= current_cpu_available:
+                logging.info(f"STARTING {s}")
+                running_futures.append(executor.submit(run_one, s))
+                current_cpu_available -= s.cpus_needed
+                submitted_suites.append(s)
+                logging.info(f"Something submitted: CPU available={current_cpu_available}")
+                logging.info(f"Currently running {len(running_futures)} suites")
+
+        if not running_futures and not submitted_suites:
+            assert abs(current_cpu_available - total_cpu_available) < 0.1, \
+                (f"When nothing is running, current_cpu_available should be equal to total_cpu_available. "
+                 f"({current_cpu_available} vs {total_cpu_available})")
+            logging.error(f"No progress can be made with these remaining suites: {suites_pending_submission}")
+            raise RuntimeError("Unschedulable suites")
+
+        # remove the just submitted suites from pending
+        for s in submitted_suites:
+            suites_pending_submission.remove(s)
+
+        try:
+            # wait for any one suite to complete
+            # main thread will spend most of its time here
+            completed_future = next(as_completed(running_futures))
+        except StopIteration:
+            # no more futures to wait on, start another submit/wait cycle
+            continue
+
+        # remove the completed future from running futures
+        running_futures.remove(completed_future)
```

Comment:
question: `as_completed` (above) could return more than one, right? why not iterate over all completed futures and remove from `running_futures`?

Reply:
`as_completed` yields one at a time.  It could yield one completion at time T, then yield another at T+1800.  When first completion arrives at time T, we need to utilize the freed up resources ASAP by submitting new suites to be run.  That's why we immediately continue to the next submit/wait loop.  Alternatively, we could immediate submit new tasks right here - but that would make the code much more complex and hard to reason about.

There probably are other models and patterns we can follow altogether - this is just what we went with here. ("Submit-many/wait-for-1")